File: encoder/v2/base_coder.py
from typing import List


class ArithmeticCoderBase:
    def __init__(self, num_bits: int = 32):
        if num_bits < 1:
            raise ValueError("State size out of range")

        self.num_state_bits = num_bits
        self.full_range = 1 << self.num_state_bits
        self.half_range = self.full_range >> 1
        self.quarter_range = self.half_range >> 1
        self.minimum_range = self.quarter_range + 2
        self.maximum_total = self.minimum_range
        self.state_mask = self.full_range - 1

        self.low = 0
        self.high = self.state_mask

    def update(self, frequencies: List[int], symbol: int) -> None:
        range_length = self.high - self.low + 1

        total = frequencies[-1]
        symbol_low = frequencies[symbol - 1] if symbol > 0 else 0
        symbol_high = frequencies[symbol]
        low = self.low

        self.low = low + symbol_low * range_length // total
        self.high = low + (symbol_high * range_length // total) - 1

        while ((self.low ^ self.high) & self.half_range) == 0:
            self.shift()
            self.low = ((self.low << 1) & self.state_mask)
            self.high = ((self.high << 1) & self.state_mask) | 1

        while (self.low & ~self.high & self.quarter_range) != 0:
            self.underflow()
            self.low = (self.low << 1) ^ self.half_range
            self.high = ((self.high ^ self.half_range) << 1) | self.half_range | 1

    def shift(self) -> None:
        raise NotImplementedError()

    def underflow(self) -> None:
        raise NotImplementedError()

File: v1/diploma/base_coder.cpp
#include "base_coder.h"


inline BaseCoder::BaseCoder() : BaseCoder(32) {
}

inline BaseCoder::BaseCoder(const size_t& num_bits) {
    if (num_bits < 1) {
        throw std::runtime_error("State size out of range");
    }
    num_bits_ = num_bits;
    full_range_ = 1 << num_bits_;
    half_range_ = full_range_ >> 1;
    quarter_range_ = half_range_ >> 1;
    minimum_range_ = quarter_range_ + 2;
    maximum_total_ = minimum_range_;
    state_mask_ = full_range_ - 1;
    low_ = 0;
    high_ = state_mask_;
}

inline void BaseCoder::Update(const std::vector<int> &frequencies, const unsigned char &symbol) {
    const auto range_length = high_ - low_ + 1;
    const auto total = frequencies.back();
    const auto symbol_code = static_cast<size_t>(symbol);
    const auto low = low_;

    low_ = low + symbol_code ? frequencies[symbol_code - 1] : 0 * range_length / total;
    high_ = low + (frequencies[symbol_code] * range_length / total) - 1;

    while (((low_ ^ high_) & half_range_) == 0) {
        Shift();
        low_ = ((low_ << 1) & state_mask_);
        high_ = ((high_ << 1) & state_mask_) | 1;
    }

    while ((low_ & ~high_ & quarter_range_) != 0) {
        Underflow();
        low_ = (low_ << 1) ^ half_range_;
        high_ = ((high_ ^ half_range_) << 1) | half_range_ | 1;
    }
}